### PR TITLE
Better error message for dataloader_idx arg

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/4206
Rework of: https://github.com/PyTorchLightning/pytorch-lightning/pull/5097
Provide better error message when:
* Users has forgotten to add the `dataloader_idx` when using multiple dataloaders (val/test)
* Users has forgotton to remove `dataloader_idx` when using single dataloader (val/test)

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/configuration_validator.py [hunk 1]
```diff
@@ -100,6 +102,20 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
...
+            len_loaders = 1 if not isinstance(loaders, list) else len(loaders)
+            if (len(sig_args) == 2 and len_loaders != 1):
+                raise MisconfigurationException(
+                    f'You provide multiple {stage} dataloaders, but no `dataloader_idx` argument'
+                    f' given in {step_name}.'
+                )
+            if (len(sig_args) == 3 and len_loaders < 2):
+                raise MisconfigurationException(
+                    f'You provided only a single {stage} dataloader, but have included the `dataloader_idx`'
+                    f' in the {step_name} method.'
```
Comment (awaelchli) [inline pytorch_lightning/trainer/configuration_validator.py:117]: what if someone has `dataloader_idx=0` in the signature, or just wants their module to be agnostic for as many dataloaders to validate as they want, including just one?

Comment (SkafteNicki) [reply pytorch_lightning/trainer/configuration_validator.py:117]: fair point, in that case we should probably check if the user has provided a default value and if not then raise an error

File: pytorch_lightning/trainer/configuration_validator.py [hunk 2]
```diff
@@ -100,6 +102,21 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
...
+        if has_step and has_loader:
+            sig_args = inspect.signature(getattr(model, step_name)).parameters
+            loaders = getattr(model, loader_name)()
+            len_loaders = 1 if not isinstance(loaders, list) else len(loaders)
+            if (len(sig_args) == 2 and len_loaders != 1):
+                raise MisconfigurationException(
+                    f'You provide multiple {stage} dataloaders, but no `dataloader_idx` argument'
+                    f' given in {step_name}.'
+                )
+            if (len(sig_args) == 3 and len_loaders < 2) and sig_args['dataloader_idx'].default == inspect._empty:
```
Comment (justusschock) [inline pytorch_lightning/trainer/configuration_validator.py:114]:
```suggestion
            if (len(sig_args) == 3 and len_loaders < 2) and sig_args['dataloader_idx'].default is not inspect.Parameter.empty:
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 3]
```diff
@@ -100,6 +102,22 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
             rank_zero_warn(f'you passed in a {loader_name} but have no {step_name}. Skipping {stage} loop')
         if has_step and not has_loader:
             rank_zero_warn(f'you defined a {step_name} but have no {loader_name}. Skipping {stage} loop')
+        if has_step and has_loader:
+            sig_args = inspect.signature(getattr(model, step_name)).parameters
+            loaders = getattr(model, loader_name)()
+            len_loaders = 1 if not isinstance(loaders, list) else len(loaders)
+            if (len(sig_args) == 2 and len_loaders != 1):
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:109]:
```suggestion
            if len(sig_args) == 2 and len_loaders != 1:
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 4]
```diff
@@ -100,6 +102,22 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
...
+                raise MisconfigurationException(
+                    f'You provide multiple {stage} dataloaders, but no `dataloader_idx` argument'
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:111]:
```suggestion
                    f'You provided multiple {stage} dataloaders, but no `dataloader_idx` argument'
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 5]
```diff
@@ -100,6 +102,22 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
...
+                    f' given in {step_name}.'
+                )
+            if (len(sig_args) == 3
+                and len_loaders < 2) and sig_args['dataloader_idx'].default is not inspect.Parameter.empty:
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:115]:
```suggestion
            if len(sig_args) == 3 and len_loaders < 2 and sig_args['dataloader_idx'].default is not inspect.Parameter.empty:
```

might need to split the line though

**Review by carmocca — APPROVED** [review]

**Review by carmocca — CHANGES_REQUESTED** [review]
Just noticed this needs to also check that the signature does not have `*args` with no kwargs or `**kwargs`

**Review by tchaton — APPROVED** [review]
After @carmocca comments, LGTM !

File: pytorch_lightning/trainer/configuration_validator.py [hunk 6]
```diff
@@ -100,6 +102,30 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
...
+            len_args = len(sig_args)
+            loaders = getattr(model, loader_name)()
+            len_loaders = 1 if not isinstance(loaders, list) else len(loaders)
+            if 'args' in sig_args:  # positional arg work with both single and multi dataloader
+                pass
+            if 'args' not in sig_args and 'kwargs' in sig_args:
+                raise MisconfigurationException(
+                    f'Method {step_name} does not work with single `kwargs` argument. Either change the signature'
+                    f' to `{step_name}(self, batch, batch_idx, dataloader_idx=0)`'
+                    f' or `{step_name}(self, *args, **kwargs)`'
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:116]: Why is this a limitation?

I would expect  `self, batch, batch_idx, **kwargs` to work by passing the dataloader_idx argument.

File: pytorch_lightning/trainer/configuration_validator.py [hunk 7]
```diff
@@ -100,6 +102,30 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
             rank_zero_warn(f'you passed in a {loader_name} but have no {step_name}. Skipping {stage} loop')
         if has_step and not has_loader:
             rank_zero_warn(f'you defined a {step_name} but have no {loader_name}. Skipping {stage} loop')
+        if has_step and has_loader:
+            sig_args = inspect.signature(getattr(model, step_name)).parameters
+            len_args = len(sig_args)
+            loaders = getattr(model, loader_name)()
+            len_loaders = 1 if not isinstance(loaders, list) else len(loaders)
+            if 'args' in sig_args:  # positional arg work with both single and multi dataloader
+                pass
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:111]: Did you mean?

```suggestion
                return
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 8]
```diff
@@ -100,6 +102,30 @@ def __verify_eval_loop_configuration(self, model: 'pl.LightningModule', stage: s
             rank_zero_warn(f'you passed in a {loader_name} but have no {step_name}. Skipping {stage} loop')
         if has_step and not has_loader:
             rank_zero_warn(f'you defined a {step_name} but have no {loader_name}. Skipping {stage} loop')
+        if has_step and has_loader:
+            sig_args = inspect.signature(getattr(model, step_name)).parameters
+            len_args = len(sig_args)
+            loaders = getattr(model, loader_name)()
+            len_loaders = 1 if not isinstance(loaders, list) else len(loaders)
+            if 'args' in sig_args:  # positional arg work with both single and multi dataloader
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:110]: Could early return `if kwargs in signature or args in signature`

**Comment by edenlightning** [comment]
@SkafteNicki can you fix conflicts?
